src/app/models/form.server.model.ts

```diff
@@ -208,7 +211,69 @@ const compileFormModel = (db: Mongoose): IFormModel => {
             'Check that your form is MyInfo-authenticated, is an email mode form and has 30 or fewer MyInfo fields.',
         },
       },
-      form_logics: [LogicSchema],
+      form_logics: {
+        type: [LogicSchema],
+        validate: {
+          validator(v: ILogicSchema[]) {
+            /**
+             * `this` variable points to the parent document object when we use
+             * save().
+             */
+            const form = this as unknown as IFormSchema
+
+            /**
+             * A condition object is said to be validatable if it contains the two
+             * necessary for validation: fieldType and state
+             */
+            type ValidatableCondition = IncompleteValidatableCondition & {
+              fieldType: BasicField
+            }
+
+            /**
+             * A validatable condition is incomplete if there is a possibility
+             * that its fieldType is null, which is a sign that a condition's
+             * field property references a non-existent form_field.
+             */
+            type IncompleteValidatableCondition = {
+              state: LogicConditionState
+              fieldType?: BasicField
+            }
+
+            const isAllConditionsReferenceExistingFields = (
+              conditions: IncompleteValidatableCondition[],
+            ): conditions is ValidatableCondition[] =>
+              !conditions.some((condition) => !condition.fieldType)
+
+            const conditions = v.flatMap((logic) => {
+              return logic.conditions.map<IncompleteValidatableCondition>(
+                (condition) => {
+                  const {
+                    field,
+                    state,
+                  }: { field: ObjectId; state: LogicConditionState } = condition
+                  return {
+                    state,
+                    fieldType: form.form_fields?.find(
+                      (f: IFieldSchema) =>
+                        f._id.toHexString() === field.toHexString(),
+                    )?.fieldType,
+                  }
+                },
+              )
+            })
+            if (!isAllConditionsReferenceExistingFields(conditions))
+              return false
```

Comment:
I think this is a bug actually. we don't prevent admins from saving forms with invalid logic (i.e. logic referencing non-existent fields), we merely warn them that their form has invalid logic. I think you may need to loosen the validation to accept conditions referencing non-existent fields.

Reply:
would be good to add model validation tests for this case also, it's important to make sure this doesn't regress as quite a few forms have invalid logic (we see the logs all the time)